### PR TITLE
optimize code in the scheduler plugin tainttoleration

### PR DESCRIPTION
**Pull request by hwdef**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

If `getPreScoreState` returns an error,the program should exit without doing other operations. This PR can improve the efficiency of the code when `getPreScoreState` reports an error.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go [hunk 1]
```diff
@@ -146,15 +146,18 @@ func countIntolerableTaintsPreferNoSchedule(taints []v1.Taint, tolerations []v1.
...
 	nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
 	if err != nil {
 		return 0, framework.AsStatus(fmt.Errorf("getting node %q from Snapshot: %w", nodeName, err))
 	}
 	node := nodeInfo.Node()
-
-	s, err := getPreScoreState(state)
-	if err != nil {
-		return 0, framework.AsStatus(err)
+	if node == nil {
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:159]: This is not necessary, if node is nil, we'll return error above. See
https://github.com/kubernetes/kubernetes/blob/8594df1aff9efde5f7e152b331093527eb9cf46a/pkg/scheduler/internal/cache/snapshot.go#L189-L194

Comment (hwdef) [reply pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go]: Thanks for your suggestions. I changed it.

File: pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go [hunk 2]
```diff
@@ -146,16 +146,16 @@ func countIntolerableTaintsPreferNoSchedule(taints []v1.Taint, tolerations []v1.
...
-	nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	s, err := getPreScoreState(state)
 	if err != nil {
-		return 0, framework.AsStatus(fmt.Errorf("getting node %q from Snapshot: %w", nodeName, err))
+		return 0, framework.AsStatus(err)
 	}
-	node := nodeInfo.Node()
 
-	s, err := getPreScoreState(state)
+	nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:154]: Another concern here is why do you think calling `getPreScoreState` first will be more efficient, they're all memory operations.

Comment (hwdef) [reply pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:154]: I do this for two reasons:
1. Get node is not easy to make mistakes, put the ones that are easy to make mistakes in the front, and put the ones that are not easy to report errors in the back, so as to improve the efficiency
2. Node is closer to nodeinfo, making it easier for people who read the code to understand the code

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:154]: We used to store state in `Map`, so we'll lock/unlock the map when getting/setting values. I think this is the main reason why we get node first. Currently, we replace `Map` with `sync.Map` which is more efficient in read-more-write-less scenarios, but we may still lock the map when reading in some circumstances.

So in general,  I'm not convinced this is more efficient to get state first. But yes, variables should sit near to the calling codes. Glad to hear others' advices. 🤔

Comment (hwdef) [reply pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:154]: Thanks for your suggestion. what you said makes sense. What do you think of this PR? If you think it's dispensable, I'll close it

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/tainttoleration/taint_toleration.go:154]: Prefer to close it then. ^ ^

**Pull request closed without merging**